topostats/classes.py

```diff
@@ -1125,30 +1235,214 @@ def __str__(self) -> str:
+        }
+        return pd.DataFrame([data])
+
+
+def convert_to_dict(to_convert: Any) -> Any:
+    """
+    Convert a class to a dictionary representation of itself.
+
+    Parameters
+    ----------
+    to_convert : Any
+        An object to be converted to a dictionary / dictionary item.
 
+    Returns
+    -------
+    Any
+        Input parameter as a dictionary / dictionary item.
+    """
+    if isinstance(to_convert, (int | float | str | Path | bool | np.ndarray | tuple | None)):
+        return to_convert
+    if isinstance(to_convert, list):
+        return [convert_to_dict(item) for item in to_convert]
+    if isinstance(to_convert, dict):
+        return {key: convert_to_dict(value) for key, value in to_convert.items()}
+    if hasattr(to_convert, "__dict__"):
+        return {re.sub(r"^_", "", key): convert_to_dict(value) for key, value in to_convert.__dict__.items()}
 
-def convert_to_dict(something: Any) -> dict[str, Any]:
+    return to_convert
+
+
+def prepare_data_for_df(class_object: object, mapping_name: str) -> list[dict[str, Any]]:
     """
-    Convert an a class to a dictionary representation of itself.
+    Gather required data for creating pd.DataFrames for CSVs.
+
+    A class object is given, and the values under the class name in the selected mapping are cycled through.
+    Basic data types are directly added to the resultant data structure, and attributes that're classes
+    will re-call the method with that class as the `class_object`, returning data to be added to the final structure.
+
+    If the attribute is a dictionary, the data types inside the dictionary are checked; if they're
+    class objects this method will be re-called for each object inside the dictionary, and again the result
+    will be added to the final structure (the data already there will also be duplicated so it exists on every row).
+    If the dictionary only contains basic data types, this data will be moved up one layer and directly
+    added to the final structure.
+    There is once case where the dictionary contains multiple nested dictionaries (grainstats output), this function
+    has specific code to handle this case but it is not generalised and will probably need working on.
+    (lines 1398 - 1417)
 
     Parameters
     ----------
-    something : Any
-        An object to be converted to a dictionary.
+    class_object : object
+        Class object to be iterated through to find required data.
+    mapping : str
+        key mapped to a dictionary of required data for the output, with class names as keys and lists containing strings of
+        required attributes from each class.
 
     Returns
     -------
-    dict[str, Any]
-        Dictionary representation of the object.
+    list[dict[str, Any]]
+        A list of dictionaries containing required data, ready to be converted into a pd.DataFrame.
     """
-    return dict(something.__dict__)
+    mappings = {
+        "image_statistics": {
```

Comment:
Lets drop the `_statistics` from each of these, less to be mis-spelt.

```suggestion
        "image": {
```